llama-index-integrations/llms/llama-index-llms-siliconflow/llama_index/llms/siliconflow/base.py

```diff
@@ -0,0 +1,479 @@
+)
+from llama_index.core.bridge.pydantic import Field, PrivateAttr
+from llama_index.core.constants import DEFAULT_NUM_OUTPUTS, DEFAULT_CONTEXT_WINDOW
+from llama_index.core.llms.callbacks import llm_chat_callback, llm_completion_callback
+from llama_index.core.llms.function_calling import FunctionCallingLLM
+from llama_index.core.base.llms.generic_utils import (
+    chat_to_completion_decorator,
+    achat_to_completion_decorator,
+    stream_chat_to_completion_decorator,
+    astream_chat_to_completion_decorator,
+)
+from llama_index.core.tools import ToolSelection
+
+if TYPE_CHECKING:
+    from llama_index.core.tools.types import BaseTool
+
+DEFAULT_SILICONFLOW_API_URL = "https://api.siliconflow.cn/v1/chat/completions"
+DEFAULT_REQUEST_TIMEOUT = 60
+AVAILABLE_OPTIONS = [
+    "deepseek-ai/DeepSeek-V2.5",
+    "deepseek-ai/DeepSeek-V2-Chat",
+    "deepseek-ai/DeepSeek-Coder-V2-Instruct",
+    "Qwen/Qwen2.5-72B-Instruct-128K",
+    "Qwen/Qwen2.5-72B-Instruct",
+    "Qwen/Qwen2-VL-72B-Instruct",
+    "Qwen/Qwen2.5-32B-Instruct",
+    "Qwen/Qwen2.5-14B-Instruct",
+    "Qwen/Qwen2.5-7B-Instruct",
+    "Qwen/Qwen2.5-Math-72B-Instruct",
+    "Qwen/Qwen2.5-Coder-7B-Instruct",
+    "Qwen/Qwen2-72B-Instruct",
+    "Qwen/Qwen2-7B-Instruct",
+    "Qwen/Qwen2-1.5B-Instruct",
+    "Qwen/Qwen2-57B-A14B-Instruct",
+    "TeleAI/TeleChat2",
+    "TeleAI/TeleMM",
+    "01-ai/Yi-1.5-34B-Chat-16K",
+    "01-ai/Yi-1.5-9B-Chat-16K",
+    "01-ai/Yi-1.5-6B-Chat",
+    "THUDM/chatglm3-6b",
+    "THUDM/glm-4-9b-chat",
+    "Vendor-A/Qwen/Qwen2-72B-Instruct",
+    "Vendor-A/Qwen/Qwen2.5-72B-Instruct",
+    "internlm/internlm2_5-7b-chat",
+    "internlm/internlm2_5-20b-chat",
+    "OpenGVLab/InternVL2-Llama3-76B",
+    "OpenGVLab/InternVL2-26B",
+    "nvidia/Llama-3.1-Nemotron-70B-Instruct",
+    "meta-llama/Meta-Llama-3.1-405B-Instruct",
+    "meta-llama/Meta-Llama-3.1-70B-Instruct",
+    "meta-llama/Meta-Llama-3.1-8B-Instruct",
+    "meta-llama/Meta-Llama-3-8B-Instruct",
+    "meta-llama/Meta-Llama-3-70B-Instruct",
+    "google/gemma-2-27b-it",
+    "google/gemma-2-9b-it",
+    "Pro/Qwen/Qwen2.5-7B-Instruct",
+    "Pro/Qwen/Qwen2-7B-Instruct",
+    "Pro/Qwen/Qwen2-1.5B-Instruct",
+    "Pro/Qwen/Qwen2-VL-7B-Instruct",
+    "Pro/01-ai/Yi-1.5-9B-Chat-16K",
+    "Pro/01-ai/Yi-1.5-6B-Chat",
+    "Pro/THUDM/chatglm3-6b",
+    "Pro/THUDM/glm-4-9b-chat",
+    "Pro/internlm/internlm2_5-7b-chat",
+    "Pro/OpenGVLab/InternVL2-8B",
+    "Pro/meta-llama/Meta-Llama-3-8B-Instruct",
+    "Pro/meta-llama/Meta-Llama-3.1-8B-Instruct",
+    "Pro/google/gemma-2-9b-it",
+]
+FUNCTION_CALLING_OPTIONS = [
+    "deepseek-ai/DeepSeek-V2.5",
+    "internlm/internlm2_5-20b-chat",
+    "internlm/internlm2_5-7b-chat",
+    "Pro/internlm/internlm2_5-7b-chat",
+    "Qwen/Qwen2.5-72B-Instruct",
+    "Qwen/Qwen2.5-32B-Instruct",
+    "Qwen/Qwen2.5-14B-Instruct",
+    "Qwen/Qwen2.5-7B-Instruct",
+    "Pro/Qwen/Qwen2.5-7B-Instruct",
+    "THUDM/glm-4-9b-chat",
+    "Pro/THUDM/glm-4-9b-chat",
+]
```

Comment:
Thought: I wonder if there's some API to ping to populate these lists on the fly. No big deal though

Reply:
I will check it.